### DISCUSSION TITLE
File validation with RHF + YUP

### DISCUSSION THREAD

**Discussion by denyschornyi** [react-hook-form/react-hook-form#5214] [Q&A] [answered]
**Describe the bug**
I know is a very hard case but we are currently in this situation. We are using 3 technologies: react-hook-form with yup and react-dropzone. And here is unexpected behavior, we expect that required fields 'name' and 'filesCdl' show the error message after submitting with empty fields, but they're no submit processed and no errors messages.  But after the first submit with entered data in inputs, that's start show errors message, but not all of them(not all files messages).
There is only one error message that appears after the second submit is "This field is required", others messages about file type and size don't appear, because don't work and do not validate.

**To Reproduce**

1. Go to https://codesandbox.io/s/nice-ives-dbmcj?file=/src/App.js
2. Leave fields empty and submit (will see that required error messages don't show)
3. Fill fields and submit
4. Leave empty fields and press submit (Will see those error messages appears)

**Codesandbox link (Required)**
https://codesandbox.io/s/nice-ives-dbmcj?file=/src/App.js

**Expected behavior**
See the error messages, not after the first submit

**Screenshots**
1. Try to sumbit with empty fields (No error messages appear )
<img width="556" alt="Screen Shot 2021-05-15 at 11 42 06 PM" src="https://user-images.githubusercontent.com/49405398/118378915-8c3bdd00-b5df-11eb-8815-2bb05ac737cf.png">
2. After first sumbit filling the fields and sumbit
<img width="556" alt="Screen Shot 2021-05-15 at 11 40 49 PM" src="https://user-images.githubusercontent.com/49405398/118378890-61518900-b5df-11eb-836f-a3751a71396f.png">

**Desktop (please complete the following information):**

- OS: [MacOS Big Sur]
- Browser [chrome]

**Additional context**
I'm using the latest possible version of all libraries.

**Comment by leapful** [comment, marked as answer]
@denyschornyi @bluebill1049 

This is not a bug. The input with type "file" always has value as a FileList object therefore it always passes required validation. A custom required validation should be implemented for this specific case.

**Reply by denyschornyi** [reply]
Thank you for answering. Do you know how to realize the custom required validation?

**Reply by leapful** [reply]
@denyschornyi 

You can implement it by using "test" API from yup like your current custom validation for file size and file type. The param "value" is the FileList object, just return false of "value.length" is 0.

For example:
```
yup.test('required', 'A file is required', value => {
   return value && value.length > 0;
});
```

**Reply by denyschornyi** [reply]
Thank you a lot for your answer.
One more question, if I need to validate a list of files, I need to loop through of them in this way?

```
files: yup
    .mixed()
    .test("required", "Photo is required", (value) => {
      return value && value.length > 0;
    })
    .test("type", "We only support jpeg and jpg", (value) => {
      if (value && value.length > 0) {
        Array.from(value).forEach((el) => {
          if (el.type === "image/jpeg") {
            return true;
          } else if (el.type === "image/jpg") {
            return true;
          } else {
            return false;
          }
        });
      }
    }),
```

**Reply by leapful** [reply]
@denyschornyi 

Yes, you can do that or just simplify the loop with Array.prototype.every:

```

value.every(file => file.type === 'image/jpeg' || file.type === 'image/jpg');

```

**Reply by denyschornyi** [reply]
Thank you a lot, now it works correctly

**Comment by bluebill1049** [comment]
Thank you @leapful!
